Source/DafnyStandardLibraries/examples/ConcurrentExamples.dfy

```diff
@@ -0,0 +1,185 @@
+    {
+      mutex := new Lock();
+      box := new AtomicBox(p1, 0);
+      primary := new MutableMap(p2);
+      secondaryA := new MutableMap(p2);
+      secondaryB := new MutableMap(p2);
+    }
+
+    ghost predicate Valid()
+      reads this`box, this`primary, this`secondaryA, this`secondaryB, this.box, this.primary, this.secondaryA, this.secondaryB
+    {
+      box.inv == p1 && primary.inv == p2 && secondaryA.inv == p2 && secondaryB.inv == p2
+      && box.Valid() && primary.Valid() && secondaryA.Valid() && secondaryB.Valid()
+    }
+
+    method Write(n: nat)
+      requires Valid()
+      modifies box
+      ensures Valid()
+    {
+      box.Put(n);
+    }
+
+    method Commit(slot: nat)
+      requires Valid()
+      modifies primary
+      ensures Valid()
+    {
+      var value := box.Get();
+      primary.Put(slot, value);
+    }
+
+    method Propagate(slot: nat)
+      requires Valid()
+      modifies secondaryA, secondaryB
+      ensures Valid()
+    {
+      mutex.Lock();
+      var val := primary.Get(slot);
+      if val.Some? {
+        secondaryA.Put(slot, val.value);
+      }
+      val := primary.Get(slot);
+      if val.Some? {
+        secondaryB.Put(slot, val.value);
+      }
+      mutex.Unlock();
+    }
+
+    method Read(copy: Copy, slot: nat) returns (r: Option<nat>)
+      requires Valid()
+    {
+      match copy
+      case A => r := secondaryA.Get(slot);
+      case B => r := secondaryB.Get(slot);
+    }
+  }
+
+  method {:test} TestApplication() {
+    var a := new Application();
+    a.Write(0);
+    a.Commit(0);
+    a.Propagate(0);
+    var r := a.Read(A, 0);
+    expect(r == Some(0));
+  }
+
+  method {:test} TestKeys() {
+    var mmap := new MutableMap(p2);
+    var keys := mmap.Keys();
+    expect(keys == {});
+    mmap.Put(0, 0);
+    keys := mmap.Keys();
+    expect(keys == {0});
+  }
+
+  method {:test} TestHasKey() {
+    var mmap := new MutableMap(p2);
+    var b := mmap.HasKey(0);
+    expect(!b);
+    mmap.Put(0, 0);
+    b := mmap.HasKey(0);
+    expect(b);
+  }
+
+  method {:test} TestValues() {
+    var mmap := new MutableMap(p2);
+    var values := mmap.Values();
+    expect(values == {});
+    mmap.Put(0, 0);
+    values := mmap.Values();
+    expect(values == {0});
+  }
+
+  method {:test} TestItems() {
+    var mmap := new MutableMap(p2);
+    var items := mmap.Items();
+    expect(items == {});
+    mmap.Put(0, 0);
+    items := mmap.Items();
+    expect(items == {(0, 0)});
+  }
+
+  method {:test} TestPutGet() {
+    var mmap := new MutableMap(p2);
+    mmap.Put(0, 0);
+    var v := mmap.Get(0);
+    expect(v == Some(0));
+  }
+
+  method {:test} TestRemove() {
+    var mmap := new MutableMap(p2);
+    mmap.Put(0, 0);
+    var b := mmap.HasKey(0);
+    expect(b);
+    assert mmap.inv(1,1);
+    mmap.Remove(1);
+    b := mmap.HasKey(0);
+    expect(b);
+    mmap.Remove(0);
+    b := mmap.HasKey(0);
+    expect(!b);
+  }
+
+  method {:test} TestCardinality() {
```

Comment:
Copy and paste typo I think, doesn't test Cardinality

Reply:
Yes! Thanks!